quotefault/__init__.py

```diff
@@ -215,28 +230,152 @@ def get():
-    elif metadata['speaker'] is not None:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).filter(Quote.speaker == metadata['speaker']).all()
-    else:
-        quotes = db.session.query(Quote, func.sum(Vote.direction).label('votes')).outerjoin(Vote).group_by(Quote).order_by(Quote.quote_time.desc()).all()
+
+    # Get all the quotes
+    quotes = get_quote_query(speaker = request.args.get('speaker'),
+        submitter = request.args.get('submitter')).all()
 
     #tie any votes the user has made to their uid
     user_votes = db.session.query(Vote).filter(Vote.voter == metadata['uid']).all()
 
-
     return render_template(
         'bootstrap/additional_quotes.html',
         quotes=quotes[20:],
         metadata=metadata,
         user_votes=user_votes
     )
+
+@app.route('/report/<quote_id>', methods=['POST'])
+@auth.oidc_auth
+def submit_report(quote_id):
+    """
+    Report a quote and notify EBoard/RTP
+    """
+    metadata = get_metadata()
+    existing_report = Report.query.filter(Report.reporter==metadata['uid'],
+        Report.quote_id==quote_id).first()
+    if existing_report:
+        flash("You already submitted a report for this Quote!")
+        return redirect('/storage')
+    new_report = Report(quote_id, metadata['uid'], "Report Reason Not Given")
+    db.session.add(new_report)
+    db.session.commit()
+    if app.config['MAIL_SERVER'] != '':
+        send_report_email( metadata['uid'], Quote.query.get(quote_id) )
+    flash("Report Successful!")
+    return redirect('/storage')
+
+@app.route('/review', methods=['GET'])
+@auth.oidc_auth
+def review():
+    """
+    Presents page for admins to review reports
+    and either hide or keep quote
+    """
+    metadata = get_metadata()
+    if metadata['is_admin']:
+        # Get the most recent 20 quotes
+        reports = Report.query.filter(Report.reviewed == False).all()
+
+        return render_template(
+            'bootstrap/admin.html',
+            reports=reports,
+            metadata=metadata
+        )
+    abort(403)
+
+@app.route('/review/<report_id>/<result>', methods=['POST'])
+@auth.oidc_auth
+def review_submit(report_id, result):
+    """
+    Called when Admin decides on a quote being hidden or kept
+    """
+    metadata = get_metadata()
+    report = Report.query.get(report_id)
+    if not metadata['is_admin']:
+        abort(403)
+    if not report:
+        abort(404)
+    if result == "keep": 
+        report.reviewed = True
+        db.session.commit()
+        flash("Report Completed: Quote Kept")
+    elif result == "hide":
+        report.quote.hidden = True
+        report.reviewed = True
+        db.session.commit()
+        flash("Report Completed: Quote Hidden")
+    else:
+        abort(404)

```

Comment:
```suggestion
        abort(400)
```
 404 is not found, 400 is bad request. Prefer 400 for invalid data, like a bad arg, and 404 for missing records, like "there's no report by that id"

Reply:
Didn't you tell me to make it 404 over 400?

Reply:
https://github.com/ComputerScienceHouse/Quotefault/pull/60#discussion_r748648468

Reply:
Oh, nvm. I got it